compiler/typecheck/TcDaml.hs

```diff
@@ -0,0 +1,37 @@
+{-# LANGUAGE CPP #-}
+
+module TcDaml(
+        customDamlErrors
+  ) where
+
+import GhcPrelude
+
+import InstEnv
+import Outputable
+import TcRnTypes
+import TyCoRep
+import TyCon
+import OccName
+
+customDamlErrors :: Ct -> [ClsInst] -> SDoc -> Maybe SDoc
+customDamlErrors ct candidate_insts binds_msg
+  | TyConApp con [TyConApp target [], viewType] <- ctev_pred (ctEvidence ct)
+  , "HasInterfaceView" <- occNameString $ occName $ tyConName con
+  = Just
+  $ vcat [ text "Tried to get an interface view of type" <+> ppr viewType <+> text "from a non-interface" <+> ppr (tyConName target)
+         , text "If" <+> ppr (tyConName target) <+> text "is a template, try casting it using toInterface or toInterfaceContractId"
+         ]
+  | TyConApp con [TyConApp target [], TyConApp choiceName [], result] <- ctev_pred (ctEvidence ct)
+  , "HasExercise" <- occNameString $ occName $ tyConName con
+  = Just
+  $ vcat [ text "Tried to exercise a choice" <+> ppr choiceName <+> text "which doesn't exist on" <+> ppr (tyConName target)
+         , text "If the choice" <+> ppr choiceName <+> text "belongs to an interface, try casting" <+> ppr (tyConName target) <+> text "using toInterface or toInterfaceContractId"
+         ]
+  | TyConApp con [TyConApp target [], LitTy (StrTyLit methodName), result] <- ctev_pred (ctEvidence ct)
+  , "HasMethod" <- occNameString $ occName $ tyConName con
+  = Just
+  $ vcat [ text "Tried to call method" <+> ppr methodName <+> text "which doesn't exist on" <+> ppr (tyConName target)
+         , text "If the method" <+> ppr methodName <+> text "belongs to an interface, try casting" <+> ppr (tyConName target) <+> text "using toInterface or toInterfaceContractId"
+         ]
```

Comment:
the `No instance for (DA.Internal.Desugar.HasMethod Interface "method" ReturnType)` error only occurs when a user tries to implement a method that isn't part of the interface, so the error should look more like this

```suggestion
  $ text "Tried to implement method" <+> ppr methodName <> text ", but interface" <+> ppr (tyConName target) <+> text "does not have a method with that name."
```

Reply:
There's a similar error message when a method is implemented with the wrong return type,

```
  • Couldn't match type ‘MyReturnType’ with ‘ReturnType’
  arising from a functional dependency between:
  constraint ‘DA.Internal.Desugar.HasMethod Interface "method" MyReturnType’
  arising from a use of ‘DA.Internal.Desugar.mkMethod’
  instance ‘DA.Internal.Desugar.HasMethod Interface "method" ReturnType’
  at ./Interface.daml:9:3-7
```

I'm not sure what's the best place to pick this one, though.

Reply:
Fixed